### PR TITLE
SQLite client

### LINKED ISSUE #1739: FEAT:  Migrate SQLite to a class-based database driver
Take the `BasicDatabaseClient` from the ultimate edition and use it to reimplement the SQLite driver.

Why?

- Much easier to manage classes vs objects
- better template for future integrations
- Typescript amirite?

### PR DESCRIPTION
**Pull request by not-night-but**
Close #1739

This PR refactors the sqlite client to be a proper class :grinning:
It also simplifies the query execution methods.

### CODE REVIEW

**Review by rathboma — CHANGES_REQUESTED** [review]
Awesome work Day! Great stuff. Only minor changes needed:

Let's not use `as` and just have two methods. Also - remove the stuff you told yourself to remove :-)

You have a bunch of notes about broken or incomplete things. Can they be dealt with or better notes left?

File: apps/studio/src/lib/db/clients/sqlite.ts [hunk 1]
```diff
@@ -0,0 +1,635 @@
...
+      table
+    }))
+  }
+
+  listSchemas(_db: string, _filter?: SchemaFilterOptions): Promise<string[]> {
+    return Promise.resolve([]); // DOES NOT SUPPORT IT
+  }
+
+  getTableReferences(_table: string, _schema?: string): Promise<string[]> {
+    return Promise.resolve([]); // TODO: not implemented yet
```
[the remark this answers is not in the document]
Comment (not-night-but) [reply apps/studio/src/lib/db/clients/sqlite.ts:207]: That was the same thing in the old driver

File: apps/studio/src/lib/db/clients/sqlite.ts [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+      toTable: row.table,
+      fromTable: table,
+      fromColumn: row.from,
+      toColumn: row.to,
+      onUpdate: row.on_update,
+      onDelete: row.on_delete
+    }))
+  }
+
+  // NOTE (@day): This may need some more work
```
[the remark this answers is not in the document]
Comment (not-night-but) [reply apps/studio/src/lib/db/clients/sqlite.ts:225]: I thought I removed this note lol, is very old

File: apps/studio/src/lib/db/clients/sqlite.ts [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+    return [];
+  }
+
+  createDatabase(databaseName: string, _charset: string, _collation: string): void {
+    // because this is a convenience for an otherwise ez-pz action, the location of the db file will be in the same location as the other .db files.
+    // If the desire for a "but I want this in another directory" is ever wanted, it can be included but for now this feels like it suits the current needs.
+    const fileLocation = this.database.split('/');
+    fileLocation.pop();
+
+    const db = new Database(`${fileLocation.join('/')}/${databaseName}.db`)
```
[the remark this answers is not in the document]
Comment (not-night-but) [reply apps/studio/src/lib/db/clients/sqlite.ts:499]: Good catch. I always forget windows likes to be different

File: apps/studio/src/lib/db/clients/sqlite/SqliteCursor.ts [hunk 4]
```diff
@@ -1,5 +1,7 @@
 import { BeeCursor } from "../../models";
 import Sqlite, { Database, Statement } from 'better-sqlite3'
+
+// TODO (@day): remove this interface, I think it might be useless now
```
[the remark this answers is not in the document]
Comment (not-night-but) [reply apps/studio/src/lib/db/clients/sqlite/SqliteCursor.ts:4]: Dope

File: apps/studio/src/lib/db/clients/sqlite.ts [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+    db.close()
+  }
+
+  createDatabaseSQL(): string {
+    throw new Error("Method not implemented.");
```
[the remark this answers is not in the document]
Comment (not-night-but) [reply apps/studio/src/lib/db/clients/sqlite.ts:504]: I think that's a good idea. Would be super useful for the readonly clients as well

File: apps/studio/src/lib/db/clients/sqlite.ts [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+
+  async listTables(_db?: string, _filter?: FilterOptions): Promise<TableOrView[]> {
+    const sql = `
+      SELECT name
+      FROM sqlite_master
+      WHERE type='table'
+      ORDER BY name
+    `;
+
+    const { data } = await this.driverExecuteQuery(sql) as SqliteResult;
```
[the remark this answers is not in the document]
Comment (not-night-but) [reply apps/studio/src/lib/db/clients/sqlite.ts:110]: Alrighty, I'll just modify the base class

**Review by rathboma — APPROVED** [review]
Looks good. Let's merge it.

**Pull request merged**